### PR TITLE
Support type hints on with and for statements

### PR DESCRIPTION
**Pull request by afrieder**
Addresses #892

### CODE REVIEW

**Comment by ddfisher** [comment]
I didn't get a chance to take a look at this today, but I should have time on Monday.

File: test-data/unit/check-statements.test [hunk 1]
```diff
@@ -1193,6 +1229,86 @@ with A() as (a, b):
...
+    def __exit__(self, x, y, z): pass
+
+class B:
+    def __enter__(self) -> str: pass
+    def __exit__(self, x, y, z): pass
+
+with A() as a, A() as (b, c), B() as d:  # type: Tuple[int, int], (int, int), str
+    pass
+
+with A() as e, A() as (f, g), B() as h:  # type: Tuple[int, int], Tuple[int, int], str
```
Comment (ddfisher) [inline test-data/unit/check-statements.test]: It's a bit unfortunate that we allow multiple ways of expressing the same thing here, but it looks like we do the same for tuple assignments (and this is really just based on how `anal_type` is written), so I'm fine with this for now.

File: mypy/fastparse.py [hunk 2]
```diff
@@ -472,10 +472,15 @@ def visit_AugAssign(self, n: ast35.AugAssign) -> OperatorAssignmentStmt:
     # For(expr target, expr iter, stmt* body, stmt* orelse, string? type_comment)
     @with_line
     def visit_For(self, n: ast35.For) -> ForStmt:
+        if n.type_comment is not None:
+            target_type = parse_type_comment(n.type_comment, n.lineno, self.errors)
```
Comment (ddfisher) [inline mypy/fastparse.py]: Nit: this should probably be called `index_type`.

File: mypy/fastparse2.py [hunk 3]
```diff
@@ -464,10 +464,15 @@ def visit_AugAssign(self, n: ast27.AugAssign) -> OperatorAssignmentStmt:
     # For(expr target, expr iter, stmt* body, stmt* orelse, string? type_comment)
     @with_line
     def visit_For(self, n: ast27.For) -> ForStmt:
+        if n.type_comment is not None:
+            target_type = parse_type_comment(n.type_comment, n.lineno, self.errors)
```
Comment (ddfisher) [inline mypy/fastparse2.py]: Nit: this should probably be called `index_type`.

**Comment by gvanrossum** [comment]
For standard unpacking mypy already supports both forms, so I think it makes sense to support the same here. E.g.
```
(a, b), c = (1, 1), "" # type: (int, int), str
(a, b), c = (1, 1), "" # type: (int, int), str
```
are both supported.

File: mypy/semanal.py [hunk 4]
```diff
@@ -2132,10 +2136,48 @@ def analyze_try_stmt(self, s: TryStmt, visitor: NodeVisitor,
...
+            elif len(actual_targets) == 1:
+                # We have one target and one type
+                types = [s.target_type]
+            elif isinstance(s.target_type, TupleType):
+                # We have multiple targets and multiple types
+                if len(actual_targets) == len(s.target_type.items):
+                    types = s.target_type.items
+                else:
+                    # But it's the wrong number of items
+                    self.fail('Incompatible number of tuple items', s)
```
Comment (ddfisher) [inline mypy/semanal.py]: I think a slightly more descriptive error message would be good here.  Maybe something like "Incompatible number of with targets and with target types"?

File: mypy/semanal.py [hunk 5]
```diff
@@ -2132,10 +2136,48 @@ def analyze_try_stmt(self, s: TryStmt, visitor: NodeVisitor,
...
+            else:
+                # We have multiple targets and one type
+                self.fail('Tuple type expected for multiple variables', s)
```
Comment (ddfisher) [inline mypy/semanal.py]: This error message is also a bit unclear.  (I don't think most users will understand what you mean by "tuple type" here.)  I'd consider something like "Single type given for multiple with targets" or "Multiple with targets require multiple types".

File: mypy/semanal.py (hunk 4 above)
Comment (ddfisher) [inline mypy/semanal.py]: (Upon further inspection, I see that you got this error message from the tuple assignment error message.  There's a decent amount of code in mypy that I would consider "historical", and therefore not the best candidate for emulation.  Not sure there's necessarily a great way for you to know which code that is until it comes up in review, though.)

File: test-data/unit/check-statements.test (hunk 1 above)
Comment (ddfisher) [inline test-data/unit/check-statements.test]: @gvanrossum: It looks like both forms are supported.  (See: this test.)

File: mypy/fastparse.py (hunk 2 above)
Comment (afrieder) [inline mypy/fastparse.py]: This block refers to `index` as `target`, so I was just consistent. Having `target` and `target_type` become `index` and `index_type` is clearer than having `target` and `index_type` become `index` and `index_type`.

File: mypy/semanal.py (hunk 4 above)
Comment (afrieder) [inline mypy/semanal.py]: So it turns out those error messages don't actually do anything. These errors are raised elsewhere, earlier. These cases are actually unreachable (but not evident by types). How should I handle this?

File: mypy/fastparse.py (hunk 2 above)
Comment (ddfisher) [inline mypy/fastparse.py]: Ah, right -- the Python AST calls it `target`.  Yeah, this is better. 👍

File: mypy/semanal.py (hunk 4 above)
Comment (ddfisher) [inline mypy/semanal.py]: These error cases are reachable when the number of targets differs from the number of elements in the top-level tuple:
```python
from typing import *

class A:
    def __enter__(self) -> Tuple[int, int]: pass
    def __exit__(self, x, y, z): pass

with A() as (a, b), A() as (c, d):  # type: int, int, int
    pass
```
(You'll have to change the error messages so they're not identical to the errors in `store_declared_types` in order to verify this.)

Comment (ddfisher) [inline mypy/semanal.py]: For that reason, I think it makes sense to change the error messages as I suggested.  The error messages in `store_declared_types` should probably be updated to be a bit better too, but that's a separate issue.

**Comment by afrieder** [comment]
@ddfisher Updated the error messages.

**Comment by ddfisher** [comment]
Thanks!

**Pull request merged**